### DISCUSSION TITLE
Thanks

### DISCUSSION THREAD

**Discussion by asad-awadia** [shihuili1218/klein#62] [General]
This project looks amazing - thanks! Will keep an eye out how it turns out

Why paxos and not raft?

**Comment by shihuili1218** [comment]
It's a good question, and one that must be answered.

Paxos was chosen because the out-of-order negotiation and Leader-Less features of Paxos are attractive.

- Out-of-order negotiation allows Paxos to use all CPU resources, whereas Raft single-threaded negotiation has a lot of wasted CPU resources.
- Leader-Less, so that Paxos has better availability.

Although they introduce some problems (e.g., live locks), there are other ways to get around them.
Introduce the Master to reduce the probability of live locks, but the Master is not Leader-Based, and Paxos can still perform write requests during the Master election.
